test_utils.go

```diff
@@ -247,6 +264,7 @@ func (a TestClient) LongCallResult(ctx context.Context) (ret int, err error) {
 
 type mockCodec struct {
 	elems []interface{}
+	mtx   sync.Mutex
```

Comment:
I thought of doing that, and in fact had it pretty much implemented, but then you hit the problem that you need to put the element back if it's the wrong type. Then you need to guard two resources instead of just one.